src/block_storage/protocol.rs

```diff
@@ -0,0 +1,201 @@
+
+#![allow(non_snake_case)]
+#![allow(missing_docs)]
+
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+protocol_enum! {
+    #[doc = "Possible volume statuses."]
+    enum VolumeStatus {
+        Creating = "creating",
+        Available = "available",
+        Reserved = "reserved",
+        Attaching = "attaching",
+        Detaching = "detaching",
+        InUse = "in-use",
+        Maintenance = "maintenance",
+        Deleting = "deleting",
+        AwaitingTransfer = "awaiting-transfer",
+        Error = "error",
+        ErrorDeleting = "error_deleting",
+        BackingUp = "backing-up",
+        RestoringBackup = "restoring-backup",
+        ErrorBackingUp = "error_backing-up",
+        ErrorRestoring = "error_restoring",
+        ErrorExtending = "error_extending",
+        Downloading = "downloading",
+        Uploading = "uploading",
+        Retyping = "retyping",
+        Extending = "extending"
+    }
+}
+
+protocol_enum! {
+    #[doc = "Available sort keys."]
+    enum VolumeSortKey {
+        CreatedAt = "created_at",
+        Id = "id",
+        Name = "name",
+        UpdatedAt = "updated_at"
+    }
+}
+
+impl Default for VolumeSortKey {
+    fn default() -> VolumeSortKey {
+        VolumeSortKey::CreatedAt
+    }
+}
+
+/// A volume attachment.
+#[derive(Debug, Clone, Deserialize)]
+pub struct VolumeAttachment {
+    pub server_id: String, // this should be a reference to a server
+    pub attachment_id: String,
+    pub attached_at: String,
+    pub host_name: String,
+    pub volume_id: String, // this should be a reference to a volume
+    pub device: String,
+    pub id: String,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct Link {
+    pub rel: String,
+    pub href: String,
+}
+
+/// A volume.
+#[derive(Debug, Clone, Deserialize)]
+pub struct Volume {
+    // TODO: not all fields fully match the API spec:
+    // https://docs.openstack.org/api-ref/block-storage/v3/#list-accessible-volumes-with-details
+    // Some fields are not actually optional, but don't work without Option<>.
+    // Others should maybe be enums, but the possible values are not documented.
+    // There are comments for these cases.
+    pub migration_status: Option<String>, // consider enum
+    pub attachments: Vec<VolumeAttachment>,
+    pub links: Vec<Link>,
+    pub availability_zone: Option<String>,
+    #[serde(rename = "os-vol-host-attr:host")]
+    pub host: Option<String>,
+    pub encrypted: bool,
+    pub encryption_key_id: Option<String>,
+    pub updated_at: Option<String>,
+    pub replication_status: Option<String>, // not optional in spec, also consider enum
+    pub snapshot_id: Option<String>,
+    pub id: String,
+    pub size: u64,
+    pub user_id: String,
+    #[serde(rename = "os-vol-tenant-attr:tenant_id")]
+    pub tenant_id: Option<String>,
+    #[serde(rename = "os-vol-mig-status-attr:migstat")]
+    pub migstat: Option<String>, // consider enum
```

Comment:
++ to a enum. We can add it later, of course, but that will be a breaking change (not the end of the world though).

In any case, could you rename it to migration_status or whatever fits? We're aiming for user-friendly names.

Reply:
Well, the reason I didn't use `migration_status` is because there already is a fields named `migration_status`. Any other suggestions? :D 

As for the enum, I will probably have to dig deeper then the docs here, because there is no table or list of valid entries.

Reply:
> Well, the reason I didn't use migration_status is because there already is a fields named migration_status. Any other suggestions? :D

Ehhhm, great :D What does this field actually do? If we don't know and cannot figure out, maybe skip it?

Reply:
Alright, I commented out this field and left a note that this is skipped due to the unclear use and uninformative naming.